harness/determined/common/api/_util.py

```diff
@@ -110,6 +111,26 @@ def read_paginated(
         offset = pagination.endIndex
 
 
+def read_paginated_with_session(
```

Comment:
suggestion: add an optional `session` to `read_paginated`?

Reply:
Actually... I don't get why this was necessary to do at all. `get_with_offset` has a `session` already referenced. Why would we need do give it a different one rather than the one that tends to be in scope when it's defined?

Reply:
this is partially because i initially implemented it in `read_paginated` but in a later commit decided to separate it out, and in doing so, left out the code that actually mattered 🙃 i fixed it, so that now `read_paginated_with_session` is responsible for the session context, and `get_with_offset` uses the session it's given.

Reply:
> Why would we need do give it a different one rather than the one that tends to be in scope when it's defined?
because the scope is different. `get_with_offset` is a singular API call that can't manage a persistent session. imagine:
```
with self._session as sess:
    def get_with_offset(int):
        return bindings.get_Trials(
            session=sess
        )
    # if we call read_paginated here, it doesn't seem right that it doesn't take in a session
    resp = read_paginated(get_with_offset)

# if we call it here outside the session context, the session context is useless
resp = read_paginated(get_with_offset)
```
it makes sense to me that `read_paginated` has the responsibility of entering/exiting the session context since it knows the "context" of requesting multiple pages in succession. then `get_with_offset` just uses the session `read_paginated_with_session` gives it. 

i agree it's not great, but given the way `read_paginated` is implemented with the separate `get_with_offset`, this is the cleanest way i could find to do it

Reply:
I think I need it explained to me what makes this a good idea.

The implicit-session interface to `read_paginated` looks to me like:
1. construct a method that maps an int to a list of things starting at that number
2. exercise the method to get a complete list of things

The explicit-session interface is:
1. construct a method that uses a session to map an int to a list of things starting at that number
2. If the passed Session has an existing HTTP connection, close it
3. Create a new persistent HTTP connection in the session passed to `read_paginated`
4. use the modified Session to call the master and get your complete list
5. close the persistent HTTP connection in the passed Session

Modifying the Session seems bad to me. If it turns out we need an explicit-session interface, then I think it should either rely on persistent-HTTP management from the caller _or_ create a brand-new session. Relying on the caller to manage it seems like the more natural interface from the point of view of a caller for the given signature.

Your intuition and mine differ about how 
```python
resp = read_paginated(get_with_offset)
``` 
looks without a passed session. I think `read_paginated`/`get_with_offset` interface is bad, but given the relationship between them, it feels more natural to me for the session to be implicit in `get_with_offset`. Whether that's closed with `read_paginated` or closed by the caller of `read_paginated`, I don't so much care.

I don't think it's necessary to support the use case in the example above. I don't think a caller should expect that that use case works.

Reply:
removed additional changes for `read_paginated`